dot/sync/block_importer.go

```diff
@@ -104,20 +104,7 @@ func (b *blockImporter) importBlock(bd *types.BlockData, origin BlockOrigin) (im
 // or the index of the block data that errored on failure.
 func (b *blockImporter) processBlockData(blockData types.BlockData, origin BlockOrigin) error {
 	if blockData.Header != nil {
-		var (
-			hasJustification = blockData.Justification != nil && len(*blockData.Justification) > 0
-			round            uint64
-			setID            uint64
-		)
-
-		if hasJustification {
-			var err error
-			round, setID, err = b.finalityGadget.VerifyBlockJustification(
-				blockData.Header.Hash(), blockData.Header.Number, *blockData.Justification)
-			if err != nil {
-				return fmt.Errorf("verifying justification: %w", err)
-			}
-		}
+		header := blockData.Header
 
 		if blockData.Body != nil {
 			err := b.processBlockDataWithHeaderAndBody(blockData, origin)
```

Comment:
mmm you mean we would have to split the `processBlockDataWithHeaderAndBody` function so one part will be executed here and the other one after verifying the justification? Or I'm missing something? 
My changes are based on a "revert" of [these changes](https://github.com/ChainSafe/gossamer/commit/1836e44e018d32da11e4b84f779418a83406f835#diff-fa462c9d158b0ade886d250e613472bb634287b6fb2025c0d818942b0ced5763)

Reply:
That will not solve the problem, because you're trying to apply the forced change before we know about it. What happens is: Forced changes are in the block digests, basically the infos inside it is the block number where this change will take place and the new auths (normaly the block number is the same as the block that contains the forced change) so when we import the block we extract the digest and get the forced change and we place it in a [`orderedPendingChanges`](https://github.com/ChainSafe/gossamer/blob/a46ad52411cb95dad9e706c8c725f00ff940f36d/dot/state/grandpa_changes.go#L38) that stores the changes in order so whenever we import a block we check if there is a change to do.

So, if you try to apply a forced change before adding it to the `orderedPendingChanges` it will never found it and will always return nil. 

Maybe you will need to open the block digest and check if there is a forced change that will take place in the current block, if so you apply it directly (following the same rules as `ApplyForcedChanges`), I believe that code should be under `dot/state/grandpa.go`